src/vs/platform/clipboard/browser/clipboardService.ts

```diff
@@ -4,15 +4,83 @@
  *--------------------------------------------------------------------------------------------*/
 
 import { $ } from 'vs/base/browser/dom';
+import Severity from 'vs/base/common/severity';
 import { URI } from 'vs/base/common/uri';
+import { localize } from 'vs/nls';
 import { IClipboardService } from 'vs/platform/clipboard/common/clipboardService';
+import { IDialogService } from 'vs/platform/dialogs/common/dialogs';
+import { IEnvironmentService } from 'vs/platform/environment/common/environment';
+import { registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { ILogService } from 'vs/platform/log/common/log';
+import { IOpenerService } from 'vs/platform/opener/common/opener';
 
 export class BrowserClipboardService implements IClipboardService {
 
+	constructor(
+		@IDialogService private readonly dialogService: IDialogService,
+		@IOpenerService private readonly openerService: IOpenerService,
+		@ILogService private readonly logService: ILogService,
+		@IEnvironmentService private readonly environmentService: IEnvironmentService,
+	) { }
+
 	declare readonly _serviceBrand: undefined;
 
 	private readonly mapTextToType = new Map<string, string>(); // unsupported in web (only in-memory)
 
+	// Guard access to navigator.clipboard with try/catch
+	// as we have seen DOMExceptions in certain browsers
+	// due to security policies. For example, in Safari,
+	// it has the following message:
+	//
+	// "The request to write to the clipboard must be triggered during a user gesture.
+	// A call to clipboard.write or clipboard.writeText outside the scope of a user
+	// gesture(such as "click" or "touch" event handlers) will result in the immediate
+	// rejection of the promise returned by the API call."
+	// From: https://webkit.org/blog/10855/async-clipboard-api/
+	//
+	// A similar limitation is there for Read. This is why this function later calls for a modal to be shown.
+	protected async doClipboardAction(action: 'write', text: string): Promise<void>;
+	protected async doClipboardAction(action: 'read'): Promise<string>;
+	protected async doClipboardAction(action: 'write' | 'read', text?: string) {
+		const isWrite = action === 'write';
+
+		try {
+			return isWrite ? await navigator.clipboard.writeText(text!) : await navigator.clipboard.readText();
+		} catch (error) {
+			// do not ask for input in tests (https://github.com/microsoft/vscode/issues/112264)
+			if (!!this.environmentService.extensionTestsLocationURI) {
```

Comment:
Actually I am quite annoyed by the fact that our integration tests write to clipboard (`paragraph 5`). I would suggest to only use the in-memory map when we detect a test run so that the clipboard is not altered.

Reply:
I would suggest something else here: instead of taking a dependency on the environment service (which is also not fulfillable for the standalone case) we could put a testing clipboard service into the DI container when running tests. That keeps the service simple (and reusable) and tests never pollute the clipboard.